diff --git a/dist/vite-plugin-config-alias/index.js b/dist/vite-plugin-config-alias/index.js
index e6cf6cc5a18d941c518d016f00f1c80e3d67813a..c5576cc8d3ce49b97aefa6361d6a792b5fd1e8ed 100644
--- a/dist/vite-plugin-config-alias/index.js
+++ b/dist/vite-plugin-config-alias/index.js
@@ -68,6 +68,7 @@ function configAliasVitePlugin({
           let hasReplacement = false;
           const replaceAliases = (match, importId) => {
             if (!importId) return match;
+            if (importId.startsWith("data:")) return match;
             const resolved = resolveWithAlias(importId, configAlias);
             if (resolved) {
               hasReplacement = true;
